fix(react-query): allow retryOnMount when throwOnError is function (#9336)

Han5991 opened the pull request:
## Summary

  - Fixed throwOnError retry logic to properly handle function vs boolean values
  - When throwOnError is a function, allow retryOnMount to proceed even when error boundary hasn't been reset
  - Maintains existing prevention behavior for boolean throwOnError values
  - Added comprehensive test coverage for the retry behavior

Han5991 commented on packages/react-query/src/errorBoundaryUtils.ts:
```diff
@@ -28,13 +28,16 @@ export const ensurePreventErrorBoundaryRetry = <
       options.retryOnMount = false
     }
+  } else if (options.throwOnError) {
+    if (!errorResetBoundary.isReset() && typeof options.throwOnError !== 'function') {
```
Updated to evaluate the throwOnError function result instead of just checking if it's a function. Thanks for the feedback!

TkDodo commented on packages/react-query/src/errorBoundaryUtils.ts:
I think we can simplify this quite a bit:

```
const throwOnError =
  query.state.error && typeof options.throwOnError === 'function'
    ? shouldThrowError(options.throwOnError, [query.state.error, query])
    : options.throwOnError
if (
  options.suspense ||
  options.experimental_prefetchInRender ||
  throwOnError
) {
  // Prevent retrying failed query if the error boundary has not been reset yet
  if (!errorResetBoundary.isReset()) {
    options.retryOnMount = false
  }
}
```

TkDodo commented on packages/react-query/src/errorBoundaryUtils.ts:
```diff
@@ -25,16 +25,24 @@ export const ensurePreventErrorBoundaryRetry = <
     TQueryKey
   >,
   errorResetBoundary: QueryErrorResetBoundaryValue,
+  query?: Query<TQueryFnData, TError, TQueryData, TQueryKey>,
```
can we make this required?

Han5991 commented:
Hi @TkDodo, thank you for the review! I've applied your suggestions:
- Moved the query retrieval after _experimental_beforeQuery in useBaseQuery.
- Simplified ensurePreventErrorBoundaryRetry logic and made the query parameter required as requested.
- Updated useQueries to safely handle the required query parameter.
All tests are passing. Ready for another look! 🙏

TkDodo commented on packages/react-query/src/errorBoundaryUtils.ts:
```diff
@@ -25,11 +25,17 @@ export const ensurePreventErrorBoundaryRetry = <
     TQueryKey
   >,
   errorResetBoundary: QueryErrorResetBoundaryValue,
+  query: Query<TQueryFnData, TError, TQueryData, TQueryKey>,
```
just seeing this now: to be on par with `getHasError`, we should make this param required, but potentially undefined, and then do a single check in here. sorry about changing my mind here 😅 


```suggestion
  query: Query<TQueryFnData, TError, TQueryData, TQueryKey> | undefined,
```

TkDodo commented on packages/react-query/src/useQueries.ts:
```diff
@@ -242,9 +242,12 @@ export function useQueries<
+    const query = client.getQueryCache().get(queryOptions.queryHash)
+    if (query) {
+      ensurePreventErrorBoundaryRetry(queryOptions, errorResetBoundary, query)
+    }
```
same here:

```suggestion
    const query = client.getQueryCache().get(queryOptions.queryHash)
    ensurePreventErrorBoundaryRetry(queryOptions, errorResetBoundary, query)
```
